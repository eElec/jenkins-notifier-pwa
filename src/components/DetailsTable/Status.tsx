import React from 'react';
import { Theme } from '@mui/system';
import styled from '@emotion/styled';
import { IJobStatus } from '../../db/types';
import { css, keyframes } from '@emotion/react';

const glow = (color: string) => keyframes`
	0% { box-shadow:0 0 4px 1px ${color} }
	50% { box-shadow:0 0 4px 3px ${color} }
	100% { box-shadow:0 0 4px 1px ${color} }
`;

type Props = { building?: boolean; theme?: Theme; status?: IJobStatus };

const Status = styled('span')(({ building, status, theme }: Props) => {
	if (theme === undefined) return;
	let palette = theme.palette.success;
	switch (status) {
		case 'SUCCESS':
			palette = theme.palette.success;
			break;
		case 'FAILURE':
			palette = theme.palette.error;
			break;
		case 'UNSTABLE':
			palette = theme.palette.warning;
			break;
		default:
			palette = {
				main: theme.palette.grey[300],
				dark: theme.palette.grey[400],
			};
	}

	return css`
		margin-right: 8px;
		height: 10px;
		width: 10px;
		border-radius: 50%;
		display: inline-block;
		background-color: ${palette.main};
		box-shadow: 0px 0px 8px ${palette.main};
		${building &&
		css`
			animation: ${glow(palette.dark)} 1s ease infinite;
		`}
	`;
});

export default Status;
