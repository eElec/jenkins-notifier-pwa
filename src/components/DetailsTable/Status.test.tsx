import { describe, expect, it } from 'vitest';
import { render } from '@test/util';
import Status from './Status';

describe('Status Component', () => {
	describe('render each status', () => {
		it('success', () => {
			const success = render(<Status status="SUCCESS" />).baseElement;
			expect(success).toMatchSnapshot();
		});

		it('failure', () => {
			const failure = render(<Status status="FAILURE" />).baseElement;
			expect(failure).toMatchSnapshot();
		});

		it('unstable', () => {
			const unstable = render(<Status status="UNSTABLE" />).baseElement;
			expect(unstable).toMatchSnapshot();
		});

		it('no status', () => {
			const result = render(<Status />).baseElement;
			expect(result).toMatchSnapshot();
		});
	});

	describe('render each building status', () => {
		it('success', () => {
			const success = render(<Status status="SUCCESS" building />).baseElement;
			expect(success).toMatchSnapshot();
		});

		it('failure', () => {
			const failure = render(<Status status="FAILURE" building />).baseElement;
			expect(failure).toMatchSnapshot();
		});

		it('unstable', () => {
			const unstable = render(
				<Status status="UNSTABLE" building />
			).baseElement;
			expect(unstable).toMatchSnapshot();
		});

		it('no status', () => {
			const result = render(<Status building />).baseElement;
			expect(result).toMatchSnapshot();
		});
	});
});
