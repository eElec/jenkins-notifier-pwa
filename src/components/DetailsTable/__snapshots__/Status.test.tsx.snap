// Vitest Snapshot v1

exports[`Status Component > render each building status > failure 1`] = `
@keyframes animation-0 {
  0% {
    box-shadow: 0 0 4px 1px rgb(178, 48, 70);
  }

  50% {
    box-shadow: 0 0 4px 3px rgb(178, 48, 70);
  }

  100% {
    box-shadow: 0 0 4px 1px rgb(178, 48, 70);
  }
}

.emotion-0 {
  margin-right: 8px;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  display: inline-block;
  background-color: #FF4565;
  box-shadow: 0px 0px 8px #FF4565;
  -webkit-animation: animation-0 1s ease infinite;
  animation: animation-0 1s ease infinite;
}

<body>
  <div>
    <span
      class="emotion-0 emotion-1"
    />
  </div>
</body>
`;

exports[`Status Component > render each building status > no status 1`] = `
@keyframes animation-0 {
  0% {
    box-shadow: 0 0 4px 1px #bdbdbd;
  }

  50% {
    box-shadow: 0 0 4px 3px #bdbdbd;
  }

  100% {
    box-shadow: 0 0 4px 1px #bdbdbd;
  }
}

.emotion-0 {
  margin-right: 8px;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  display: inline-block;
  background-color: #e0e0e0;
  box-shadow: 0px 0px 8px #e0e0e0;
  -webkit-animation: animation-0 1s ease infinite;
  animation: animation-0 1s ease infinite;
}

<body>
  <div>
    <span
      class="emotion-0 emotion-1"
    />
  </div>
</body>
`;

exports[`Status Component > render each building status > success 1`] = `
@keyframes animation-0 {
  0% {
    box-shadow: 0 0 4px 1px rgb(88, 175, 78);
  }

  50% {
    box-shadow: 0 0 4px 3px rgb(88, 175, 78);
  }

  100% {
    box-shadow: 0 0 4px 1px rgb(88, 175, 78);
  }
}

.emotion-0 {
  margin-right: 8px;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  display: inline-block;
  background-color: #7FFA70;
  box-shadow: 0px 0px 8px #7FFA70;
  -webkit-animation: animation-0 1s ease infinite;
  animation: animation-0 1s ease infinite;
}

<body>
  <div>
    <span
      class="emotion-0 emotion-1"
    />
  </div>
</body>
`;

exports[`Status Component > render each building status > unstable 1`] = `
@keyframes animation-0 {
  0% {
    box-shadow: 0 0 4px 1px rgb(175, 144, 87);
  }

  50% {
    box-shadow: 0 0 4px 3px rgb(175, 144, 87);
  }

  100% {
    box-shadow: 0 0 4px 1px rgb(175, 144, 87);
  }
}

.emotion-0 {
  margin-right: 8px;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  display: inline-block;
  background-color: #FACE7D;
  box-shadow: 0px 0px 8px #FACE7D;
  -webkit-animation: animation-0 1s ease infinite;
  animation: animation-0 1s ease infinite;
}

<body>
  <div>
    <span
      class="emotion-0 emotion-1"
    />
  </div>
</body>
`;

exports[`Status Component > render each status > failure 1`] = `
.emotion-0 {
  margin-right: 8px;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  display: inline-block;
  background-color: #FF4565;
  box-shadow: 0px 0px 8px #FF4565;
}

<body>
  <div>
    <span
      class="emotion-0 emotion-1"
    />
  </div>
</body>
`;

exports[`Status Component > render each status > no status 1`] = `
.emotion-0 {
  margin-right: 8px;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  display: inline-block;
  background-color: #e0e0e0;
  box-shadow: 0px 0px 8px #e0e0e0;
}

<body>
  <div>
    <span
      class="emotion-0 emotion-1"
    />
  </div>
</body>
`;

exports[`Status Component > render each status > success 1`] = `
.emotion-0 {
  margin-right: 8px;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  display: inline-block;
  background-color: #7FFA70;
  box-shadow: 0px 0px 8px #7FFA70;
}

<body>
  <div>
    <span
      class="emotion-0 emotion-1"
    />
  </div>
</body>
`;

exports[`Status Component > render each status > unstable 1`] = `
.emotion-0 {
  margin-right: 8px;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  display: inline-block;
  background-color: #FACE7D;
  box-shadow: 0px 0px 8px #FACE7D;
}

<body>
  <div>
    <span
      class="emotion-0 emotion-1"
    />
  </div>
</body>
`;
